File: src/processing.py
import os
from time import time
import logging

from .database import update_sequences, get_sequences_from_file


def process_batch(sequences: list[str]):
    grouped_sequences: list[list[str]] = [[] for _ in range(26)]

    for seq in sequences:
        seq = seq.lower()
        index = ord(seq[1]) - 97
        if index < 0 or index >= 26:
            logging.error(f"Invalid sequence {seq}. Skipping.")
            continue
        grouped_sequences[ord(seq[1]) - 97].append(seq)

    for batch in grouped_sequences:
        if batch:
            logging.info(f"Processing batch {batch[0][:2]} ({len(batch)} sequences)")
            batch_start = time()
            update_sequences(batch)
            logging.info(f"Batch {batch[0][:2]} processed in {time() - batch_start:.2f} seconds")


def process_letter(sequences: list[str]):
    logging.info("\n--- Processing sequences ---")
    start_time = time()

    process_batch(sequences)
    logging.info(f"--- Completed in {time() - start_time:.2f} seconds ---\n")


def process_single_file(file_path: str):
    """Process a single file, splitting sequences by letter and then processing them."""
    file_name = os.path.basename(file_path)
    logging.info(f"\n--- Processing single file: {file_name} ---")

    sequences = get_sequences_from_file(file_path)
    if not sequences:
        logging.info("No sequences found. Skipping.\n")
        return

    grouped_sequences: list[list[str]] = [[] for _ in range(26)]
    for seq in sequences:
        seq = seq.lower()
        grouped_sequences[ord(seq[0]) - 97].append(seq)

    for batch in grouped_sequences:
        if batch:
            logging.info(f"Processing batch ({len(batch)} sequences)")
            process_letter(batch)

File: src/processing.ts
import { basename } from "node:path";
import { getSequencesFromFile, markSequencesAsFound } from "./database";

function processBatch(sequences: string[]) {
  const groupedSequences: string[][] = Array.from({ length: 26 }, () => []);

  for (const seq of sequences) {
    const loweredSeq = seq.toLowerCase();
    const index = loweredSeq.charCodeAt(1) - 97;
    if (index < 0 || index >= 26) {
      console.error(`Invalid sequence ${loweredSeq}. Skipping.`);
      continue;
    }
    groupedSequences[index].push(loweredSeq);
  }

  for (const batch of groupedSequences) {
    if (batch.length > 0) {
      console.info(
        `Processing batch ${batch[0].slice(0, 2)} (${batch.length} sequences)`
      );
      const batchStart = Date.now();
      markSequencesAsFound(batch);
      console.info(
        `Batch ${batch[0].slice(0, 2)} processed in ${(
          (Date.now() - batchStart) /
          1000
        ).toFixed(2)} seconds`
      );
    }
  }
}

function processLetter(sequences: string[]) {
  console.info("\n--- Processing sequences ---");
  const startTime = Date.now();

  processBatch(sequences);
  console.info(
    `--- Completed in ${((Date.now() - startTime) / 1000).toFixed(
      2
    )} seconds ---\n`
  );
}

/**
 * Process a single file, splitting sequences by letter and then processing them.
 */
export async function processSingleFile(filePath: string) {
  const fileName = basename(filePath);
  console.info(`\n--- Processing single file: ${fileName} ---`);

  const sequences = await getSequencesFromFile(filePath);
  if (sequences.length === 0) {
    console.info("No sequences found. Skipping.\n");
    return;
  }

  // Split sequences by letter
  const groupedSequences: string[][] = Array.from({ length: 26 }, () => []);
  for (const seq of sequences) {
    const loweredSeq = seq.toLowerCase(); // Ensure sequence is lowercase
    const index = loweredSeq.charCodeAt(0) - 97;
    if (index >= 0 && index < 26) {
      groupedSequences[index].push(loweredSeq);
    }
  }

  // Now process each batch of sequences for each letter
  for (const batch of groupedSequences) {
    if (batch.length > 0) {
      console.info(`Processing batch (${batch.length} sequences)`);
      processLetter(batch);
    }
  }
}
